cpp/include/milvus-storage/common/lrucache.h

```diff
@@ -0,0 +1,127 @@
+// Copyright 2023 Zilliz
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#pragma once
+
+#include <memory>
+#include <string>
+#include <mutex>
+#include <map>
+#include <shared_mutex>
+#include <functional>
+#include <list>
+
+#include <arrow/status.h>
+#include <arrow/result.h>
+
+namespace milvus_storage {
+
+template <typename K, typename V>
+class LRUCache {
+  public:
+  static LRUCache& getInstance() {
+    static LRUCache instance;
+    return instance;
+  }
+
+  // Set the capacity of cache
+  void set_capacity(size_t capacity) {
+    std::unique_lock lock(mutex_);
+    capacity_ = capacity;
+    // evict if current size exceeds new capacity
+    while (cache_.size() > capacity_) {
+      auto last_key = lru_list_.back();
+      cache_.erase(last_key);
+      lru_list_.pop_back();
+    }
+  }
+
+  // Get the value associated with the key
+  [[nodiscard]] arrow::Result<V> get(const K& key, std::function<arrow::Result<V>(const K&)> CreateFunc) {
+    // fast path: check under shared lock
+    {
+      std::unique_lock lock(mutex_);
+      auto it = cache_.find(key);
```

Comment:
unique_lock or shared_lock?

Reply:
it's actully the write lock here, because we may update the LRU list if current key hit.

```
lru_list_.splice(lru_list_.begin(), lru_list_, it2->second.second);`
```